dbms/src/Storages/DeltaMerge/DeltaMergeStore.cpp

```diff
@@ -946,6 +946,116 @@ BlockInputStreams DeltaMergeStore::readRaw(const Context & db_context,
+    const Context & db_context,
+    const DB::Settings & db_settings,
+    const ColumnDefines & columns_to_read,
+    size_t num_streams,
+    bool keep_order,
+    const SegmentIdSet & read_segments,
+    size_t extra_table_id_index)
+{
+    SegmentReadTasks tasks;
+
+    auto dm_context = newDMContext(db_context, db_settings, fmt::format("read_raw_{}", db_context.getCurrentQueryId()));
+    // If keep order is required, disable read thread.
+    auto enable_read_thread = db_context.getSettingsRef().dt_enable_read_thread && !keep_order;
+    {
+        std::shared_lock lock(read_write_mutex);
+
+        for (const auto & [handle, segment] : segments)
+        {
+            (void)handle;
+            if (read_segments.empty() || read_segments.count(segment->segmentId()))
+            {
+                auto segment_snap = segment->createSnapshot(*dm_context, false, CurrentMetrics::DT_SnapshotOfReadRaw);
+                if (unlikely(!segment_snap))
+                    throw Exception("Failed to get segment snap", ErrorCodes::LOGICAL_ERROR);
+
+                tasks.push_back(std::make_shared<SegmentReadTask>(segment, segment_snap, RowKeyRanges{segment->getRowKeyRange()}));
+            }
+        }
+    }
+
+    fiu_do_on(FailPoints::force_slow_page_storage_snapshot_release, {
+        std::thread thread_hold_snapshots([this, tasks]() {
+            LOG_WARNING(log, "failpoint force_slow_page_storage_snapshot_release begin");
+            std::this_thread::sleep_for(std::chrono::seconds(5 * 60));
+            (void)tasks;
+            LOG_WARNING(log, "failpoint force_slow_page_storage_snapshot_release end");
+        });
+        thread_hold_snapshots.detach();
+    });
+
+    auto after_segment_read = [&](const DMContextPtr & dm_context_, const SegmentPtr & segment_) {
+        this->checkSegmentUpdate(dm_context_, segment_, ThreadType::Read);
+    };
+    size_t final_num_stream = enable_read_thread
+        ? std::max(1, num_streams)
+        : std::max(1, std::min(num_streams, tasks.size()));
+    String req_info;
+    if (db_context.getDAGContext() != nullptr && db_context.getDAGContext()->isMPPTask())
+        req_info = db_context.getDAGContext()->getMPPTaskId().toString();
+    auto read_task_pool = std::make_shared<SegmentReadTaskPool>(
+        physical_table_id,
+        extra_table_id_index,
+        dm_context,
+        columns_to_read,
+        EMPTY_FILTER,
+        std::numeric_limits<UInt64>::max(),
+        DEFAULT_BLOCK_SIZE,
+        /* read_mode */ ReadMode::Raw,
+        std::move(tasks),
+        after_segment_read,
+        req_info,
+        enable_read_thread,
+        final_num_stream);
+
+    if (enable_read_thread)
+    {
+        for (size_t i = 0; i < final_num_stream; ++i)
+        {
+            group_builder.addConcurrency(
+                std::make_unique<UnorderedSourceOp>(
+                    exec_context,
+                    read_task_pool,
+                    columns_to_read,
```

Comment:
```suggestion
                    filter && filter->extra_cast ? *filter->columns_after_cast : columns_to_read,
```

Reply:
Seems useless because filter is empty in `readRaw`.